File: biddo_client/lib/core/models/category.dart
import 'package:biddo/utils/constants.dart';
import 'package:get/get.dart';

import 'asset.dart';

class Category {
  String id;
  Map<String, String> name;
  String? icon;
  String? remoteIconUrl;
  String? assetId;
  Asset? asset;

  String? parentCategoryId;
  int auctionsCount;

  List<Rx<Category>> subcategories = [];

  Category({
    required this.id,
    required this.name,
    this.icon,
    this.remoteIconUrl,
    this.assetId,
    this.asset,
    this.auctionsCount = 0,
    this.parentCategoryId,
    this.subcategories = const [],
  });

  static Category fromJSON(dynamic data) {
    return Category(
        id: data['id'],
        auctionsCount: data['auctionsCount'] != null
            ? int.parse(data['auctionsCount'].toString())
            : 0,
        icon: data['icon'] ?? '',
        remoteIconUrl: data['remoteIconUrl'] ?? '',
        assetId: data['assetId'] ?? '',
        parentCategoryId: data['parentCategoryId'],
        subcategories: data['subcategories'] != null
            ? (data['subcategories'] as List)
                .map((e) => Category.fromJSON(e).obs)
                .toList()
            : [],
        asset: data['asset'] != null ? Asset.fromJSON(data['asset']) : null,
        name: Map.fromEntries(
          Constants.LANGUAGES.map((lang) => MapEntry(
                lang['code']!,
                data['name'][lang['code']] ?? '',
              )),
        ));
  }

  Map<String, dynamic> toJSON() {
    return {
      'id': id,
      'name': name,
      'auctionsCount': auctionsCount,
      'parentCategoryId': parentCategoryId,
      'remoteIconUrl': remoteIconUrl,
      'assetId': assetId,
      'icon': icon,
      'subcategories': subcategories.map((e) => e.value.toJSON()).toList(),
      'asset': asset?.toJson(),
    };
  }
}

File: core/domain/category.ts
import { Asset } from './asset'

export class Category {
  id: string
  name: Record<string, string>
  icon: string
  remoteIconUrl?: string
  assetId?: string
  asset?: Asset

  parentCategoryId?: string
  auctionsCount?: number
  subcategories?: Category[]

  createdAt?: Date

  constructor(params: Record<string, unknown> = {}) {
    this.id = params.id as string
    this.name = params.name as Record<string, string>
    this.icon = params.icon as string
    this.remoteIconUrl = params.remoteIconUrl as string
    this.assetId = params.assetId as string
    this.asset = params.asset as Asset
    this.subcategories = params.subcategories as Category[]
    this.parentCategoryId = params.parentCategoryId as string
    this.createdAt = params.createdAt as Date
    this.auctionsCount = params.auctionsCount ? parseInt(params.auctionsCount.toString()) : 0
  }

  static fromJSON(params: Record<string, unknown> = {}) {
    const subcategories: Category[] = params.subcategories
      ? (params.subcategories as Record<string, unknown>[]).map((subcategory) =>
          Category.fromJSON(subcategory)
        )
      : []

    return new Category({
      subcategories,
      id: params.id,
      name: params.name,
      icon: params.icon,
      remoteIconUrl: params.remoteIconUrl,
      assetId: params.assetId,
      parentCategoryId: params.parentCategoryId,
      asset: params.asset ? Asset.fromJSON(params.asset as Record<string, unknown>) : undefined,
      createdAt: params.createdAt ? new Date(params.createdAt.toString()) : undefined,
      auctionsCount: params.auctionsCount ? parseInt(params.auctionsCount.toString()) : 0,
      unclassifiedGenre: params.unclassifiedGenre,
    })
  }
}
